File: Prefix Sum/백준3020.cpp
//13:48시작
#include <iostream>
using namespace std;
#define INF 987654321
int N, H;
int sum[500001], up[500002], down[500002]; //최종 구간합 구할 배열, 위에서 내려오는애 체크할 배열, 아래에서 올라온애 체크할 배열
int _min = INF, mincnt = 0;
void input()
{
	cin >> N >> H;
	for(int i = 0; i < N; i++)
	{
		int num;
		cin >> num;

		if (i % 2 == 0)
			down[num] ++;
		else
			up[H + 1 - num] ++;
	}
}
void solveUp()
{
	for (int i = H; i >= 1; i--)
	{
		down[i] += down[i + 1];
		sum[i] += down[i];
	}
}
void solveDown()
{
	for (int i = 1; i <= H; i++)
	{
		up[i] += up[i - 1];
		sum[i] += up[i];
	}
}
void getAnswer()
{
	for (int i = 1; i <= H; i++)
	{
		if (sum[i] < _min)
		{
			_min = sum[i];
			mincnt = 1;
		}
		else if (sum[i] == _min)
		{
			mincnt++;
		}
		
	}
}
int main()
{
	input();
	
	solveUp();
	solveDown();
	/*cout << "배열보여주기" << endl;
	for (int i = H; i >= 1; i--)
	{
		cout << sum[i] << endl;
	}
	cout << "배열보여주기" << endl;*/
	getAnswer();
	cout << _min << " " << mincnt << endl;
	return 0;
}

File: 이분탐색/백준3020.py
import sys
def lower_bound(arr, key):
    start = 0
    end = len(arr)
    while start < end:
        mid = (start + end) // 2
        if arr[mid] >= key:
            end = mid
        elif arr[mid] < key:
            start = mid + 1
    return end

#입력 부분
input_list = list(input().split())
N = int(input_list[0])
H = int(input_list[1])
down = []
up = []
for idx in range(N):
    input_list2 = sys.stdin.readline().split()
    num = int(input_list2[0])
    if idx % 2 == 0:
        down.append(num)
    else:
        up.append(num)

#solve
#getanswer
_min = 987654321
mincnt = 0

down.sort()
up.sort()
for idx in range(1, H + 1):
    downcnt = ( (N//2) -  lower_bound(down, idx) )
    upcnt =  ( (N//2) - lower_bound(up, H + 1 - idx) )
    if downcnt + upcnt < _min:
        _min = downcnt + upcnt
        mincnt = 1
    elif downcnt + upcnt == _min:
        mincnt = mincnt + 1


print(_min, mincnt)